### ISSUE TITLE
IRIG B002 Unmodulated Decoding

### ISSUE THREAD

**Issue by dayweiss728** [joshuaferrara/IRIG-Arduino#1]
Hi Joshua, 

My name is Sam Day-Weiss and I am an undergraduate working in a lab at Yale University. I have been working with your arduino script for the last couple of weeks and am running into some issues, specifically with the calculation of the year. Within the run() function at the beginning of IRIGB.cpp, I am struggling to see how the pos value (initially pos = -1) is incremented to gather each set of bits. It appears as though only bits at pos = 4 are pushed to the Queue to be analyzed. Even with this, I don't see how the code even understands pos to be equal to 4 without any incrementation. On a similarly difficult note, I am misunderstanding exactly how the queueSum() function is actually calculating numerical values for BCD of 7,6, and 10 bits, as opposed to 8 which is the standard BCD bit number (even with 8, the number of bits for year, it is not entirely clear how this works). Can you help me with this? All other elements besides the calculation of year seem to be working fine. 

Thanks so much,

Sam

**Comment by joshuaferrara** [comment]
Will get back to you this weekend.

**Comment by joshuaferrara** [comment]
Alright, I'll try my best to explain it. I've gone into some detail, so feel free to skip over stuff if it's unnecessary, but hopefully it'll help others in the future as well. 

Unfortunately I won't be able to do any testing to reproduce the issue, so I hope this explanation can help w.r.t. getting it working for your use case. I've included some links in the explanation that'll reference back to parts of the code I'm referring to. I'll also reference some stuff found in the [IRIG time code format specification](https://apps.dtic.mil/dtic/tr/fulltext/u2/1013738.pdf). IIRC, this was only tested with IRIG B000, not B002. 

**Note:** as I was writing this, I came upon [this image](https://www.meinbergglobal.com/images/grafik/irig-b.jpg) on [this page](https://www.meinbergglobal.com/english/info/irig.htm) which seems to say that B002 format does not include the year. So that might be the issue, but nevertheless I've kept the code explanation below.

---

There are roughly 2 states to the decoder:
1) `WAIT` - Waiting for start of frame marker (`pos = -1`)
2) `RX` - Receiving data (`pos >= 0`).

So, `pos` is initially set to `-1` until [we receive](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L28-L30) our two start-of-frame position markers, which in code are referred to as `'R'`. This is the `WAIT` state. Ideally, we should jump back to this state when we receive weird or invalid data, but I never had the need for this.

![image](https://user-images.githubusercontent.com/248680/54090434-65aff680-4331-11e9-9023-11e4a6549044.png)

(Image from IRIG spec linked above, page 36. I pretty much used this image specifically when implementing the code, so it might be a good visual to help explain it.)

Once `pos` is equal to 0, we're currently in a sort of [recording state](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L32-L38) and are [pushing the bits we receive onto the queue](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L37). When we're in the `RX` state, and we receive a position marker, we [increment pos by 1](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L34) and call `parseBits` which will dump the queue and sum up everything in it (`queueSum`). If incrementing pos by 1 happens to be equal to 9 - our last position marker - we reset it back to 0.

For example, the queue should end up looking like this when we have received all 8 bits of seconds data for a value of 55 seconds:

```
mag: 1 2 4 8 0 10 20 40
bit: 1 0 1 0 0  1  0  1
```

We then [pop everything off one-by-one](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L125), multiply the bit with the corresponding value in the `pVal` array at the same "index" of the value on the queue, and then add it up. `queueSum` will essentially evaluate this expression:

`(1*1) + (2*0) + (4*1) + (8*0) + (0*0) + (10*1) + (20*0) + (40*1) = 55`

Then, depending on the position markers we're between, assign the value we summed up to the [corresponding class member variable](https://github.com/joshuaferrara/IRIG-Arduino/blob/master/IRIGB.cpp#L147). **Note:** I'm pretty sure I was off-by-one somewhere, as all of my calls to `queueSum` tend to have their length argument set to one more than necessary.

We do all of this until we receive two position markers again, resetting the `pos` value back to 0 and starting the process again.

**Comment by joshuaferrara** [comment]
> On a similarly difficult note, I am misunderstanding exactly how the queueSum() function is actually calculating numerical values for BCD of 7,6, and 10 bits, as opposed to 8 which is the standard BCD bit number (even with 8, the number of bits for year, it is not entirely clear how this works). Can you help me with this? All other elements besides the calculation of year seem to be working fine.

Oh also - I don't know if the explanation above helps with this, but if it doesn't, could you clarify what you mean by this?

**Comment by dayweiss728** [comment]
Thank you so much for this, your explanation helps a lot. I actually discovered that it was the satellite data itself that was seemingly outputting all zeros between P5 and P6 which should encode information for year. I fiddled around with the timecard settings and got it to do what it was supposed to and the code worked beautifully. The combination of doing this and your explanation has answered all of my questions, thank you so much again for putting time into the response!

Also you're not off by one, the extra number in your queueSum() arguments correspond to a reference between the 1 2 4 8 and the 10 20 40 80 sections. These are break points between subwords that have the same length as the 0 bit, and so they need to be taken into account as well in the addition with pVal[].
